docs: Update dependencies for documentation build system (Sphinx, add-ons etc.)

qmonnet opened the pull request:
The objectives of this PR are:

- Synchronising the rstcheck version between the minimal list of requirements and the list of actual dependencies for the docs build system,
- Making sure that the Google Analytics add-on is in that minimal list so that it doesn't disappear when regenerating the full list,
- Making sure reports from the new version of rstcheck are addressed,
- Updating (`make update-requirements`) the other dependencies, as much as we can.
- In particular, making sure we update markdown-it-py, for which [Dependabot reported there is a CVE](https://github.com/cilium/cilium/pull/23980) on our current version, even though I don't think we're affected.
- Remove the config for Dependabot, so it stops offering updates on Documentation/requirements-min/requirements.txt, to avoid getting out of sync again in the future.

Please refer to individual commits for details.

A new docs-builder image will be necessary, or the CI workflow will fail because of the old rstcheck not recognising the new syntax for its arguments.

qmonnet committed:
docs: Update build dependencies (Sphinx add-ons etc.)

Update all that we can in
Documentation/requirements-min/requirements.txt, then run the command
"make update-requirements" to update all dependencies accordingly.

There are some dependencies that we cannot update to their latest
version:

- Sphinx itself cannot be updated to 6+, because myth-parser does not
  support it.

- sphinx-tabs cannot be updated from 3.4.0 to 3.4.1, because the latter
  needs docutils 0.18 which is not currently supported by our
  sphinx_rtd_theme fork (it would need a rebase).

Signed-off-by: Quentin Monnet <quentin@isovalent.com>

qmonnet committed:
docs: Update rstcheck

Dependabot has updated the version for rstcheck in
Documentation/requirements-min/requirements.txt, but it has never been
synchronised with Documentation/requirements.txt after that.

And as it happens, rstcheck 6.1.1 has had several breaking changes since
3.3.1, and a few updates are necessary:

- Option --report has been replaced with --report-level.

- Option --ignore-language has been replaced with --ignore-languages.

- Option --ignore-message has been replaced with --ignore-messages.

- For some reason, the ..openapi directive is now reported as an issue.

- It seems that rstcheck is better at analysing code blocks, but this
  results in many reports for our C blocks (which are not meant to be
  parsed as standalone snippets), so we skip analysis of C blocks.

- But the issue with malformed tables has been fixed and we can remove
  the ignore on the relevant message.

At last, the most annoying thing is that this version has a bug, showing
"critical" level messages about AttributeError reports. It mentions that
it may be related to code blocks without a language specified, but we
don't have these (or are not supposed to), and other people have
reported this might also be caused by directives in included text
fragments. See the relevant GitHub issue for more details:
https://github.com/rstcheck/rstcheck-core/issues/3. At least, these
messages do not cause rstcheck to return with an error, so we can deal
with them by simply filtering them out.

New reports from rstcheck 6.1.1 that we do not want to ignore have been
fixed in previous commits.

Fixes: #22155
Fixes: a819d979d231 ("build(deps): bump rstcheck in /Documentation/requirements-min")
Signed-off-by: Quentin Monnet <quentin@isovalent.com>

qmonnet committed:
docs: Do not print error when spell check output dir is empty

To make sure there are no spelling errors when building the docs, we
look at the output directory for the spell checker, with "ls". But if
the directory has not been created, for example because no spelling
mistake has been found, this may pring an error in the user's terminal.

Make sure we discard any error from this "ls" command.

Signed-off-by: Quentin Monnet <quentin@isovalent.com>

qmonnet committed:
docs: Fix indentation of a table in CI docs

The table should be indented with the item list it belongs to, or Sphinx
will believe the list is finished and open a new one, under the table,
starting from 3.

Signed-off-by: Quentin Monnet <quentin@isovalent.com>

qmonnet committed:
docs: Fix double reference in benchmark docs

There are two target references (non-anonymous links) names "netperf",
and pointing to the same link anyway. Let's clean it up.

Signed-off-by: Quentin Monnet <quentin@isovalent.com>

qmonnet committed:
docs: Add Google Analytics add-on to minimal list of requirements

The Sphinx plug-in for Google Analytics has been added to the full list
of runtime dependencies for the build system for the documentation, but
never to the subset list from which the full one is generated.

As a consequence, regenerating the full list with the related command,
"make -C Documentation update-requirements", removes this dependency.

Add the plug-in to the minimal list so that we don't lose it by
accident.

Fixes: #22821
Fixes: 4bc562961138 ("Add sphinxcontrib-googleanalytics to doc requirements")
Signed-off-by: Quentin Monnet <quentin@isovalent.com>

qmonnet committed:
dependabot: Remove doc build system updates, Dependabot config itself

Remove Dependabot's updates for required dependencies.

The reason is that when Dependabot attempts to update
Documentation/requirements-min/requirements.txt, then it is not enough
to merge this PR. Instead, we have to synchronise with the actual list
of dependencies ("make -C Documentation update-requirements"), and to
check that everything still works as intended.

Most of the time, Dependabot's updates for documentation requirements
are just noise, because the update don't bring any useful feature, and
we don't have the cycles to sync and test the changes for each PR, this
is simply too much work for too little benefit.

Worse, some PRs might be merged without taking the time to sync and
test, for example: https://github.com/cilium/cilium/pull/22155, and in
such case further updates of the full requirements list will be broken.

So let's stick to manual updates again.

Regarding security updates, it appears that Dependabot offers them
regardless of this configuration file (by updating directly
Documentation/requirements.txt), as seen in
https://github.com/cilium/cilium/pull/23980.

Signed-off-by: Quentin Monnet <quentin@isovalent.com>

joestringer reviewed (commented):
LGTM except for generating the new docs-builder image.

Seems like there's a similar issue with the docs-builder here as we have with the cilium-builder / cilium-runtime images. I agree for now it's OK to just handle this manually. If this bothers you, I would suggest looking a bit closer into (a) the way that the `images/` infrastructure detects changes and figures out canonical shas for a given set of input files, (b) the bot that can automatically build an image and update the commits in-line. Andre and I can help with any necessary docker repository configurations etc.

joestringer commented on .github/workflows/documentation.yaml:
```diff
@@ -59,7 +59,7 @@ jobs:
         run: |
           make -C Documentation copy-api # necessary for check-build.sh
-      - uses: docker://cilium/docs-builder:2023-02-04@sha256:67c383332f63708fc40f3d12b0ffff8ec94a9a83d1b819dbe23b226ef076e6f2
+      - uses: docker://qmonnet/docs-builder:2023-02-24
```
cilium/docs-builder:2023-02-24 is on the way. I think you should be able to retrieve the sha from here once it's built? https://hub.docker.com/repository/registry-1.docker.io/cilium/docs-builder/builds/a4d7572f-ce1b-414b-9243-2e28d2f64815

qmonnet committed:
ci: Update docs-builder image for CI workflow

Following the recent updates to the docs components, in particular the
version bump for rstcheck, a new docs-builder image is necessary.
Otherwise, rstcheck in the old image will not recognising the new syntax
for its arguments, and the workflow will fail.

Signed-off-by: Quentin Monnet <quentin@isovalent.com>
